secrets.go

```diff
@@ -0,0 +1,153 @@
+package pipeline
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/buildkite/go-pipeline/ordered"
+	"gopkg.in/yaml.v3"
+)
+
+var _ interface {
+	json.Unmarshaler
+	ordered.Unmarshaler
+} = (*Secrets)(nil)
+
+// Secrets is a sequence of secrets. It is useful for unmarshaling.
+type Secrets []*Secret
+
+// UnmarshalOrdered unmarshals Secrets from []any (sequence of secret names).
+func (s *Secrets) UnmarshalOrdered(o any) error {
+	switch o := o.(type) {
+	case nil:
+		// `secrets: null` is invalid - should be omitted entirely or use valid formats
+		return fmt.Errorf("unmarshaling secrets: secrets cannot be null")
+
+	case []any:
+		for _, c := range o {
+			switch ct := c.(type) {
+			case string:
+				secret := &Secret{
+					Key:                 ct,
+					EnvironmentVariable: &ct, // Default EnvironmentVariable to key value for simple string format
+				}
+				*s = append(*s, secret)
+
+			case map[string]any:
+				secret := &Secret{}
+
+				if key, ok := ct["key"].(string); ok {
+					secret.Key = key
+				}
+
+				if envVar, ok := ct["environment_variable"].(string); ok {
+					secret.EnvironmentVariable = &envVar
+				}
+
+				// Validate that we have at least a key
+				if secret.Key == "" {
+					return fmt.Errorf("unmarshaling secrets: secret object missing required 'key' field")
+				}
+
+				*s = append(*s, secret)
+
+			case *ordered.Map[string, interface{}]:
+				// Backend sends ordered.Map format
+				secret := &Secret{}
+
+				if keyVal, _ := ct.Get("key"); keyVal != nil {
+					if key, ok := keyVal.(string); ok {
+						secret.Key = key
+					}
+				}
+
+				if envVarVal, _ := ct.Get("environment_variable"); envVarVal != nil {
+					if envVar, ok := envVarVal.(string); ok {
+						secret.EnvironmentVariable = &envVar
+					}
+				}
```

Comment:
Similarly it might be better UX to report if the user wrote a non-string given YAML sometimes parses things non-obviously:


```suggestion
				if envVarVal, _ := ct.Get("environment_variable"); envVarVal != nil {
					envVar, ok := envVarVal.(string)
					if !ok {
						return fmt.Errorf("unmarshaling secret: environment_variable must be a string, but was %T", envVarVal)
					}
					secret.EnvironmentVariable = &envVar
				}		
```